src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Roslyn.Utilities;
+using SQLitePCL;
+
+namespace Microsoft.CodeAnalysis.SQLite.Interop
+{
+    /// <summary>
+    /// Encapsulates a connection to a sqlite database.  On construction an attempt will be made
+    /// to open the DB if it exists, or create it if it does not.
+    /// 
+    /// Connections are considered relatively heavyweight and are pooled until the <see cref="SQLitePersistentStorage"/>
+    /// is <see cref="SQLitePersistentStorage.Close"/>d.  Connections can be used by different threads,
+    /// but only as long as they are used by one thread at a time.  They are not safe for concurrent
+    /// use by several threads.
+    /// 
+    /// <see cref="SqlStatement"/>s can be created through the user of <see cref="GetResettableStatement"/>.
+    /// These statements are cached for the lifetime of the connection and are only finalized
+    /// (i.e. destroyed) when the connection is closed.
+    /// </summary>
+    internal partial class SqlConnection
+    {
+        /// <summary>
+        /// The raw handle to the underlying DB.
+        /// </summary>
+        private readonly sqlite3 _handle;
+
+        /// <summary>
+        /// Our cache of prepared statements for given sql strings.
+        /// </summary>
+        private readonly Dictionary<string, SqlStatement> _queryToStatement = new Dictionary<string, SqlStatement>();
+
+        /// <summary>
+        /// Whether or not we're in a transaction.
+        /// </summary>
+        private bool _inTransaction;
+
+        public SqlConnection(string databasePath)
+        {
+            var flags = OpenFlags.SQLITE_OPEN_CREATE | OpenFlags.SQLITE_OPEN_READWRITE;
+
+            var result = (Result)raw.sqlite3_open_v2(databasePath, out var handle, (int)flags, vfs: null);
+
+            if (result != Result.OK)
+            {
+                throw new SqlException(result, $"Could not open database file: {databasePath} ({result})");
+            }
+
+            Contract.ThrowIfNull(handle);
+
+            _handle = handle;
+            raw.sqlite3_busy_timeout(_handle, (int)TimeSpan.FromMinutes(1).TotalMilliseconds);
+        }
+
+        ~SqlConnection()
```

Comment:
> Failing to close the connection is a resource leak, but not an unrecoverable error.

That's why we report an error so we can discover and fix the resource leak.  If this code hits, it means invariants we've tried to set up are not being met and that has to be investigated.  We know from experience that people do not run debug builds.  So the only thing that can catch this would be debug tests.  And there are lots of issues we hit in the wild that we never hit in our debug tests.  

Making this not fire in release builds just means that we could end up with a resource leak in practice that we had no idea about.